content/blog/2019-10-08-remote-read-meets-streaming.md

```diff
@@ -0,0 +1,301 @@
+
+Prometheus since the 1.x version had the ability to interact directly with its storage using [remote API](https://prometheus.io/docs/prometheus/latest/storage/#remote-storage-integrations). 
+
+This API offers two methods: **write** and **read** which allows a 3rd party system to either:
+
+* Receive pushed samples by Prometheus: "Write"
+* Pull samples from Prometheus or to Prometheus: "Read"
+
+![Remote read and write architecture](/assets/blog/2019-10-08/remote_integrations.png)
+
+Both methods are using HTTP with messages encoded with [protobufs](https://github.com/protocolbuffers/protobuf). 
+Additionally, both request and response for both methods are optionally compressed using [snappy](https://github.com/google/snappy).
+
+### Remote Write 
+
+This is the most popular way to replicate Prometheus data into 3rd party system. In this mode, Prometheus streams samples, 
+by periodically sending a batch of those to the given endpoint. 
+
+Remote read was recently improved massively in March with [WAL-based remote read](https://grafana.com/blog/2019/03/25/whats-new-in-prometheus-2.8-wal-based-remote-write/) which
+improved the reliability and resource consumption. It is also worth to note that Write is supported by almost all 3rd 
+party integrations mentioned [here](https://prometheus.io/docs/operating/integrations/#remote-endpoints-and-storage).
+
+### Remote Read
+
+Read method is a bit less popular. It was added in [March 2017](https://github.com/prometheus/prometheus/commit/febed48703b6f82b54b4e1927c53ab6c46257c2f) (server side) and did not change much
+since back then. Prometheus 2.13.0 release fixes the known bottlenecks of the Read, so we will focus on the Read method in this article.
+
+The key idea of the remote read is to allow querying Prometheus storage, [the TSDB](https://github.com/prometheus/prometheus/tree/master/tsdb) directly without PromQL evaluation. 
+Similar interface, [`Querier`](https://github.com/prometheus/prometheus/blob/91d7175eaac18b00e370965f3a8186cc40bf9f55/storage/interface.go#L53), PromQL engine is using to get data from storage. 
+
+The remote read request is a very simple HTTP request with the following protobuf payload:
+
+```
+message Query {
+  int64 start_timestamp_ms = 1;
+  int64 end_timestamp_ms = 2;
+  repeated prometheus.LabelMatcher matchers = 3;
+  prometheus.ReadHints hints = 4;
+}
+```
+
+With that payload, client can ask for certain series matching given `matchers` and time range with `end` and `start`.
+
+The response is equally simple. It returns the matched time series with **raw** samples of value and timestamp.
+
+```
+message Sample {
+  double value    = 1;
+  int64 timestamp = 2;
+}
+
+message TimeSeries {
+  repeated Label labels   = 1;
+  repeated Sample samples = 2;
+}
+
+message QueryResult {
+  repeated prometheus.TimeSeries timeseries = 1;
+}
+```
+
+This essentially allows read access to metrics in TSDB that Prometheus collected. The main use cases for remote read are:
```

Comment:
This section describing why the remote read is necessary duplicates a section above. IMO - this should be the first thing that the user reads. Move this section above the proto description.